### PR TITLE
Improve search failure classification 

### PR DESCRIPTION
**Pull request by smalyshev**
This implements search failure classification by looking up sets of exception classes in the cause chain. 

Also adds more failure tests.

### CODE REVIEW

**Review by quux00 — COMMENTED** [review]
This is a good list! I added a few suggestions.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsage.java [hunk 1]
```diff
@@ -89,6 +106,53 @@ public Builder setRemotesCount(int remotesCount) {
...
+                return Result.SECURITY;
+            }
+            if (ExceptionsHelper.unwrap(e, SearchTimeoutException.class) != null) {
+                return Result.TIMEOUT;
+            }
+            if (e instanceof TaskCancelledException || (ExceptionsHelper.unwrap(e, TaskCancelledException.class) != null)) {
+                return Result.CANCELED;
+            }
+            if (ExceptionsHelper.unwrap(e, TransportException.class) != null) {
+                return Result.REMOTES_UNAVAILABLE;
```
[the remark this answers is not in the document]
Comment (smalyshev) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsage.java:134]: OK I'll borrow some code from `TransportResolveClusterAction` and improve the check.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsage.java [hunk 2]
```diff
@@ -89,6 +106,53 @@ public Builder setRemotesCount(int remotesCount) {
...
+         */
+        public static Result getFailureType(Exception e) {
+            var unwrapped = ExceptionsHelper.unwrapCause(e);
+            if (unwrapped instanceof Exception) {
+                e = (Exception) unwrapped;
+            }
+            if (ExceptionsHelper.unwrapCorruption(e) != null) {
+                return Result.CORRUPTION;
+            }
+            if (ExceptionsHelper.unwrap(e, ResourceNotFoundException.class) != null) {
```
[the remark this answers is not in the document]
Comment (smalyshev) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsage.java:121]: `IndexNotFoundException` is a subclass of `ResourceNotFoundException` so it should catch all descendants of `ResourceNotFoundException`.

**Review by quux00 — APPROVED** [review]
LGTM. Nice work here.

**Pull request merged**